tuftool/src/add_role.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::datetime::parse_datetime;
+use crate::error::{self, Result};
+use crate::source::parse_key_source;
+use chrono::{DateTime, Utc};
+use snafu::{OptionExt, ResultExt};
+use std::fs::File;
+use std::num::NonZeroU64;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use tempfile::tempdir;
+use tough::editor::{targets::TargetsEditor, RepositoryEditor};
+use tough::http::HttpTransport;
+use tough::key_source::KeySource;
+use tough::schema::PathSet;
+use tough::Transport;
+use tough::{ExpirationEnforcement, FilesystemTransport, Limits, Repository};
+use url::Url;
+
+#[derive(Debug, StructOpt)]
+pub(crate) struct AddRoleArgs {
+    /// The role being delegated
+    #[structopt(short = "d", long = "delegated-role")]
+    delegatee: String,
+
+    /// Key files to sign with
+    #[structopt(short = "k", long = "key", required = true, parse(try_from_str = parse_key_source))]
+    keys: Vec<Box<dyn KeySource>>,
+
+    /// Expiration of new role file; can be in full RFC 3339 format, or something like 'in
+    /// 7 days'
+    #[structopt(short = "e", long = "expires", parse(try_from_str = parse_datetime))]
+    expires: DateTime<Utc>,
+
+    /// Version of targets.json file
+    #[structopt(short = "v", long = "version")]
+    version: NonZeroU64,
+
+    /// Path to root.json file for the repository
+    #[structopt(short = "r", long = "root")]
+    root: PathBuf,
+
+    /// TUF repository metadata base URL
+    #[structopt(short = "m", long = "metadata-url")]
+    metadata_base_url: Url,
+
+    /// Incoming metadata
+    #[structopt(short = "i", long = "incoming-metadata")]
+    indir: Url,
+
+    /// threshold of signatures to sign delegatee
+    #[structopt(short = "t", long = "threshold")]
+    threshold: NonZeroU64,
+
+    /// The directory where the repository will be written
+    #[structopt(short = "o", long = "outdir")]
+    outdir: PathBuf,
+
+    /// The delegated paths
+    #[structopt(short = "p", long = "paths")]
+    paths: Option<Vec<String>>,
+
+    /// The delegated paths hash prefixes
+    #[structopt(short = "hp", long = "path-hash-prefixes")]
+    path_hash_prefixes: Option<Vec<String>>,
```

Comment:
I think these are mutually exclusive right? If so, structopt gives you the capability to ensure only one of these arguments gets passed.